flight/Modules/Actuator/triflight.c

```diff
@@ -0,0 +1,1018 @@
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+#include "openpilot.h"
+#include "triflight.h"
+#include "misc_math.h"
+#include "physical_constants.h"
+
+#define TRI_TAIL_SERVO_ANGLE_MID  90.0f
+#define TRI_TAIL_SERVO_ANGLE_MAX  50.0f
+#define TRI_YAW_FORCE_CURVE_SIZE  100
+
+static float tail_motor_acceleration_delay_angle;
+static float tail_motor_deceleration_delay_angle;
+static float tail_motor_pitch_zero_angle;
+static float tail_servo_max_yaw_force = 0.0f;
+static float yaw_force_curve[TRI_YAW_FORCE_CURVE_SIZE];
+
+/**
+ *
+ *
+ */
+float getPitchCorrectionAtTailAngle(float angle, float thrust_factor)
+{
+    return 1.0f / (sinf(angle) - cosf(angle) / thrust_factor);
+}
+
+/**
+ *
+ *
+ */
+void triflightInit(ActuatorSettingsData  *actuatorSettings,
+                   TriflightSettingsData *triflightSettings,
+                   TriflightStatusData   *triflightStatus)
+{
+	float min_angle = TRI_TAIL_SERVO_ANGLE_MID - triflightSettings->ServoMaxAngle;
+	float max_angle = TRI_TAIL_SERVO_ANGLE_MID + triflightSettings->ServoMaxAngle;
+
+	float max_neg_force = 0.0f;
+	float max_pos_force = 0.0f;
+
+	float angle = TRI_TAIL_SERVO_ANGLE_MID - TRI_TAIL_SERVO_ANGLE_MAX;
+
+	for (uint8_t i = 0; i < TRI_YAW_FORCE_CURVE_SIZE; i++)
+	{
+		float angle_rad = DEG2RAD * angle;
+
+		yaw_force_curve[i] = (-triflightSettings->MotorThrustFactor * cosf(angle_rad) - sinf(angle_rad)) *
+                              getPitchCorrectionAtTailAngle(angle_rad, triflightSettings->MotorThrustFactor);
+
+		// Only calculate the top forces in the configured angle range
+		if ((angle >= min_angle) && (angle <= max_angle))
+		{
+			max_neg_force = fminf(yaw_force_curve[i], max_neg_force);
+			max_pos_force = fmaxf(yaw_force_curve[i], max_pos_force);
+		}
+
+		angle++;
+	}
+
+	tail_servo_max_yaw_force = fminf(fabsf(max_neg_force), fabsf(max_pos_force));
+
+	tail_motor_pitch_zero_angle = 2.0f * (atanf(((sqrtf(triflightSettings->MotorThrustFactor *
+	                                                    triflightSettings->MotorThrustFactor + 1) + 1) /
+	                                                    triflightSettings->MotorThrustFactor)));
+
+	tail_motor_acceleration_delay_angle = (triflightSettings->MotorAccelDelayMs / 1000.0f) * triflightSettings->ServoSpeed;
+    tail_motor_deceleration_delay_angle = (triflightSettings->MotorDecelDelayMs / 1000.0f) * triflightSettings->ServoSpeed;
+}
+
+/**
+ *
+ *
+ */
+float getAngleFromYawCurveAtForce(float force)
+{
+	if (force < yaw_force_curve[0]) // No force that low
+	{
+		return TRI_TAIL_SERVO_ANGLE_MID - TRI_TAIL_SERVO_ANGLE_MAX;
+	}
+	else if (!(force < yaw_force_curve[TRI_YAW_FORCE_CURVE_SIZE - 1])) // No force that high
+	{
+		return TRI_TAIL_SERVO_ANGLE_MID + TRI_TAIL_SERVO_ANGLE_MAX;
+	}
+
+	// Binary search: yaw_force_curve[lower] <= force, yaw_force_curve[higher] > force
+	uint8_t lower  = 0;
+	uint8_t higher = TRI_YAW_FORCE_CURVE_SIZE - 1;
+
+	while (higher > lower + 1)
```

Comment:
(maybe should just go through it in order, actually, to simplify the code, and to eliminate the two conditionals at top).